sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java

```diff
@@ -0,0 +1,303 @@
+    long estimatedSize = initialSource.getEstimatedSizeBytes(options);
+    LOGGER.info("Estimated size: {}", estimatedSize);
+    assertTrue("Wrong estimated size", estimatedSize > 40000);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testRead() throws Exception {
+    insertTestDocuments(NB_DOCS);
+
+    TestPipeline pipeline = TestPipeline.create();
+
+    PCollection<String> output = pipeline.apply(
+        ElasticsearchIO.read().withConnectionConfiguration(connectionConfiguration)
+                              .withScrollKeepalive("5m"));
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally())).isEqualTo(NB_DOCS);
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testReadWithQuery() throws Exception {
+    insertTestDocuments(NB_DOCS);
+
+    String query = "{\n"
+        + "  \"query\": {\n"
+        + "  \"match\" : {\n"
+        + "    \"scientist\" : {\n"
+        + "      \"query\" : \"Einstein\",\n"
+        + "      \"type\" : \"boolean\"\n"
+        + "    }\n"
+        + "  }\n"
+        + "  }\n"
+        + "}";
+
+    Pipeline pipeline = TestPipeline.create();
+
+    PCollection<String> output = pipeline.apply(
+        ElasticsearchIO.read().withConnectionConfiguration(connectionConfiguration)
+                              .withQuery(query));
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally())).isEqualTo(NB_DOCS / 10);
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWrite() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+    List<String> data = createDocuments(NB_DOCS, false);
+    pipeline.apply(Create.of(data)).apply(ElasticsearchIO.write()
+      .withConnectionConfiguration(connectionConfiguration));
+    pipeline.run();
+
+    // upgrade
+    node.client().admin().indices().upgrade(new UpgradeRequest()).actionGet();
+
+    SearchResponse response = node.client().prepareSearch().execute().actionGet(5000);
+    assertEquals(NB_DOCS, response.getHits().getTotalHits());
+
+    QueryBuilder queryBuilder = QueryBuilders.queryStringQuery("Einstein").field("scientist");
+    response = node.client().prepareSearch().setQuery(queryBuilder).execute().actionGet();
+    assertEquals(NB_DOCS / 10, response.getHits().getTotalHits());
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWriteWithErrors() throws Exception {
+    TestPipeline pipeline = TestPipeline.create();
+    List<String> data = createDocuments(NB_DOCS, true);
+    pipeline.apply(Create.of(data)).apply(ElasticsearchIO.write().
+      withConnectionConfiguration(connectionConfiguration));
+    try {
+      pipeline.run();
+      fail("Exception should have been thrown");
+    } catch (Exception e) {
+        //type of e is Pipeline$PipelineExecutionException
+
+      String message = e.getCause().getMessage();
+      assertTrue("Wrong exception thrown", e.getCause().getClass() == IOException.class);
```

Comment:
ok, a standard hamcrest matcher for Exception class and a custom hamcrest matcher for regExp should to it better.